cherrypy/test/test_httputil.py

```diff
@@ -79,3 +80,84 @@ def test_invalid_status(status_code, error_msg):
+
+@pytest.mark.parametrize(
+    ('header_content', 'value', 'params'),
+    [
+        ('application/x-www-form-urlencoded',
+         'application/x-www-form-urlencoded', {}),  # default_content_type
+        ('application/json;charset="utf8"',
+         'application/json', {'charset': 'utf8'}),
+        ('audio/*; q=0.2, audio/basic',
+         'audio/*', {'q': '0.2, audio/basic'}),
+        ('text/plain; q=0.5, text/html, text/x-dvi; q=0.8, text/x-c',
+         'text/plain', {'q': '0.8, text/x-c'}),
+        ('text/*, text/html, text/html;level=1, */*',
+         'text/*, text/html, text/html', {'level': '1, */*'}),
+        ('iso-8859-5, unicode-1-1;q=0.8',
+         'iso-8859-5, unicode-1-1', {'q': '0.8'}),
+        ('gzip;q=1.0, identity; q=0.5, *;q=0',
+         'gzip', {'q': '0'}),
+        ('da, en-gb;q=0.8, en;q=0.7',
+         'da, en-gb', {'q': '0.7'}),
+        ('text/plain',
+         'text/plain', {}),
+        ('text/xml',
+         'text/xml', {}),
+    ]
+)
+def test_header_element(header_content, value, params):
+    """ Test that value and params are parsed from headers.
+
+        Testing that the value and params are parsed from
+        headers passed to HeaderElement from httputils """
+    hdr_elem = httputil.HeaderElement.from_str(header_content)
+
+    assert hdr_elem.value == value
+    assert hdr_elem.params == params
+
+
+@pytest.mark.parametrize(
+    'header_content,media_type,qvalue',
+    [
+        ('application/x-www-form-urlencoded',
+         'application/x-www-form-urlencoded', 1.0),  # default_content_type
+        ('application/json;charset="utf8"',
+         'application/json', 1.0),
+        ('text/*, text/html, text/html;level=1, */*',
+         'text/*, text/html, text/html', 1.0),
+        ('iso-8859-5, unicode-1-1;q=0.8',
+         'iso-8859-5, unicode-1-1', 0.8),
+        ('text/plain',
+         'text/plain', 1.0),
+        ('text/xml',
+         'text/xml', 1.0),
+    ]
+)
+def test_accept_element(header_content, media_type, qvalue):
+    """ Test that value and qvalue are parsed from headers passed to
+        AcceptElement from httputils """
+    acc_elem = httputil.AcceptElement.from_str(header_content)
+    assert acc_elem.value == media_type
+    assert acc_elem.qvalue == qvalue
+
+
+@pytest.mark.parametrize(
+    'header_content,media_type,qvalue',
+    [
+        ('audio/*; q=0.2, audio/basic',
+         'audio/*', {'q': '0.2, audio/basic'}),
+        ('text/plain; q=0.5, text/html, text/x-dvi; q=0.8, text/x-c',
+         'text/plain', {'q': '0.8, text/x-c'}),
+        ('gzip;q=1.0, identity; q=0.5, *;q=0',
+         'gzip', {'q': '0'}),
+        ('da, en-gb;q=0.8, en;q=0.7',
+         'da, en-gb', 0.8),
+    ]
+)
+def test_accept_element_raises_400(header_content, media_type, qvalue):
+    """ Test that headers passed to AcceptElement from httputils
+        raise an expected HTTPError """
+    httputil.AcceptElement.from_str(header_content)
+    assert pytest.raises(cherrypy.HTTPError)
```

Comment:
This is an incorrect way of using `pytest.raises()`. There are two main problems with it:
1) It's always best to use the `match=` argument to state what the expectation is, otherwise, a too broad set of exception might make it pass.
2) This helper returns a context manager, and so it's meant to be used in the `with pytest.raises(Exc, match=r'^stuff$'):` block.

I suppose, it should look something like this:
```suggestion
    with pytest.raises(
            cherrypy.HTTPError,
            match=r'^Malformed HTTP header: `[^`]+`$',
    ):
        httputil.AcceptElement.from_str(header_content)
```